var/spack/repos/builtin/packages/gaussian-view/package.py

```diff
@@ -44,6 +44,14 @@ class GaussianView(Package):
 
     conflicts('+gaussian-src', when='@:6.0.99')
 
+    depends_on('libx11', type='run')
```

Comment:
If you set

```
depends_on('libx11', type=('run', 'link'))
```

Then Spack should add RPATHs to the `gaussian-view` binaries pointing to the Spack-built `libx11` libraries - in that case the `gaussian-view` binaries should use the Spack-built `libx11` libs even if there are other libraries available on the system.

Reply:
@scheibelp by directly modifying the executables/libraries?

Reply:
> While this also work for .so (e.g. plugins)?

Yes. It should be possible to confirm this with `readelf -d` on any generated `.so`.

> Also, will spack use the libs property of the listed packages?

Yes (so for example if you update `def libs` in one of these dependencies, it will be reflected in the `-L` arguments when building the dependent).

Reply:
@scheibelp just tried with `vmd`, doesn't work:

```
$ ldd /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/vmd_LINUXAMD64
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/vmd_LINUXAMD64: /usr/lib64/libGL.so.1: no version information available (required by /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/vmd_LINUXAMD64)
	linux-vdso.so.1 (0x00007fff4b596000)
	libGL.so.1 => /usr/lib64/libGL.so.1 (0x00007f617a27f000)
	libX11.so.6 => /usr/lib64/libX11.so.6 (0x00007f6179f3e000)
	libXinerama.so.1 => /usr/lib64/libXinerama.so.1 (0x00007f6179d3b000)
	libXi.so.6 => /usr/lib64/libXi.so.6 (0x00007f6179b2a000)
	liboptixu.so.1 => /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/liboptixu.so.1 (0x00007f617974d000)
	liboptix.so.1 => /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/liboptix.so.1 (0x00007f61776c6000)
	libospray.so.0 => /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/libospray.so.0 (0x00007f6177113000)
	libospray_common.so.0 => /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/libospray_common.so.0 (0x00007f6176efe000)
	libembree.so.2 => /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/libembree.so.2 (0x00007f6175d1c000)
	libtbb.so.2 => /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/libtbb.so.2 (0x00007f6175ac6000)
	libtbbmalloc.so.2 => /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/libtbbmalloc.so.2 (0x00007f6175873000)
	libpthread.so.0 => /lib64/libpthread.so.0 (0x00007f6175654000)
	libm.so.6 => /lib64/libm.so.6 (0x00007f617531c000)
	libdl.so.2 => /lib64/libdl.so.2 (0x00007f6175118000)
	libutil.so.1 => /lib64/libutil.so.1 (0x00007f6174f15000)
	libstdc++.so.6 => /usr/lib64/libstdc++.so.6 (0x00007f6174b3b000)
	libgcc_s.so.1 => /lib64/libgcc_s.so.1 (0x00007f6174922000)
	libc.so.6 => /lib64/libc.so.6 (0x00007f6174567000)
	librt.so.1 => /lib64/librt.so.1 (0x00007f617435f000)
	libGLX.so.0 => /usr/lib64/libGLX.so.0 (0x00007f617412d000)
	libGLdispatch.so.0 => /usr/lib64/libGLdispatch.so.0 (0x00007f6173e71000)
	libxcb.so.1 => /usr/lib64/libxcb.so.1 (0x00007f6173c48000)
	libXext.so.6 => /usr/lib64/libXext.so.6 (0x00007f6173a36000)
	/lib64/ld-linux-x86-64.so.2 (0x00007f617a517000)
	libXau.so.6 => /usr/lib64/libXau.so.6 (0x00007f6173832000)
```

with:
```
    depends_on('libx11', type=('run', 'link'))
    depends_on('libxi', type=('run', 'link'))
    depends_on('libxinerama', type=('run', 'link'))
    depends_on('gl@3:', type=('run', 'link'))
```

and the OS variants are referenced.
Also, the executable was not changed compared to the original one.

Reply:
... with Spack 0.16.1-1701-0c8df39fce

Reply:
Could you show the output of `readelf -d /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/vmd_LINUXAMD64`?

Reply:
I'm wondering if the Spack `libx11`/`libxi` packages need to `def libs` to (the default might not properly locate the libraries): can you also show the output of

```
find `spack location -i libx11` -name "libX11*"
```

?

Reply:
> `readelf -d /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/vmd_LINUXAMD64`

```
> readelf -d /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/vmd-1.9.3-pxvond2pthsd3gubsorb4sqfsj22av7w/lib64/vmd_LINUXAMD64

Dynamic section at offset 0x15acb80 contains 39 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libGL.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libX11.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libXinerama.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libXi.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [liboptixu.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [liboptix.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libospray.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libospray_common.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libembree.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libtbb.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libtbbmalloc.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libutil.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x000000000000000f (RPATH)              Library rpath: [/]
 0x000000000000000c (INIT)               0x4e3c28
 0x000000000000000d (FINI)               0xd3d2e8
 0x0000000000000004 (HASH)               0x400260
 0x0000000000000005 (STRTAB)             0x468e58
 0x0000000000000006 (SYMTAB)             0x415fc8
 0x000000000000000a (STRSZ)              449744 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000015 (DEBUG)              0x0
 0x0000000000000003 (PLTGOT)             0x1bad880
 0x0000000000000002 (PLTRELSZ)           21120 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x4de9a8
 0x0000000000000007 (RELA)               0x4ddb38
 0x0000000000000008 (RELASZ)             3696 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffffe (VERNEED)            0x4dd9b8
 0x000000006fffffff (VERNEEDNUM)         8
 0x000000006ffffff0 (VERSYM)             0x4d6b28
 0x0000000000000000 (NULL)               0x0
```

Reply:
> I'm wondering if the Spack `libx11`/`libxi` packages need to `def libs` to

That's already part of this PR.

> (the default might not properly locate the libraries): can you also show the output of
> 
> ```
> find `spack location -i libx11` -name "libX11*"
> ```
> 
> ?

```
> find `spack location -i libx11` -name "libX11*"/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/share/doc/libX11
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/share/doc/libX11/libX11
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/share/doc/libX11/libX11/libX11.xml
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/share/doc/libX11/i18n/compose/libX11-keys.xml
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11.so.6.4.0
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11.so.6
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11.so
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11-xcb.so.1.0.0
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11-xcb.so.1
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11-xcb.so
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11.a
/sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/libx11-1.7.0-vjcdmapajed4qeoqhfyah3ey2mf4b5e7/lib64/libX11-xcb.a
```

Reply:
also doesn't work for gaussian-view:

```
> module load gaussian-view/6.1.6-gcc-7.5.0-dzoa2l5 
> /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/gaussian-view-6.1.6-dzoa2l5ya4i7uu5mvmqwdriyurnlutbx/gview.exe: error while loading shared libraries: libGLU.so.1: cannot open shared object file: No such file or directory
> readelf -d /sw-ng/spack/opt/spack/linux-opensuse_leap15-sandybridge/gcc-7.5.0/gaussian-view-6.1.6-dzoa2l5ya4i7uu5mvmqwdriyurnlutbx/gview.exe 

Dynamic section at offset 0x7c1140 contains 52 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libhdf5_sc.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libhdf5_hl_sc.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libnetcdf_sc.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libz_sc.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libqwtplot3d_sc.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libqwt_sc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libtiff_sc.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libjpeg_sc.so.9a]
 0x0000000000000001 (NEEDED)             Shared library: [libpng_sc.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libfreetype_sc.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libftgl_sc.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libscutil.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libscutil_gui.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libscutil_gl.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libicons_sc.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libGL.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libGLU.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libQtSvg.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libQtSql.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libQtXml.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libQtOpenGL.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libQtGui.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libQtNetwork.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libQtCore.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libXext.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libX11.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x000000000000000f (RPATH)              Library rpath: [$ORIGIN:$ORIGIN/lib]
 0x000000000000001d (RUNPATH)            Library runpath: [$ORIGIN:$ORIGIN/lib]
 0x000000000000000c (INIT)               0x465708
 0x000000000000000d (FINI)               0xa4b6a4
 0x0000000000000004 (HASH)               0x400290
 0x0000000000000005 (STRTAB)             0x41f530
 0x0000000000000006 (SYMTAB)             0x405e50
 0x000000000000000a (STRSZ)              180770 (bytes)
 0x000000000000000b (SYMENT)             24 (bytes)
 0x0000000000000015 (DEBUG)              0x0
 0x0000000000000003 (PLTGOT)             0xcc14d8
 0x0000000000000002 (PLTRELSZ)           93816 (bytes)
 0x0000000000000014 (PLTREL)             RELA
 0x0000000000000017 (JMPREL)             0x44e890
 0x0000000000000007 (RELA)               0x44d9f0
 0x0000000000000008 (RELASZ)             3744 (bytes)
 0x0000000000000009 (RELAENT)            24 (bytes)
 0x000000006ffffffe (VERNEED)            0x44d940
 0x000000006fffffff (VERNEEDNUM)         5
 0x000000006ffffff0 (VERSYM)             0x44b752
 0x0000000000000000 (NULL)               0x0
```

Reply:
I just read https://github.com/spack/spack/pull/22322#discussion_r602375816 which points out that this is a precompiled binary. This probably *should* use the `link` deptype for `libx11` etc. but that will not be sufficient

This could take a similar approach to `intel-oneapi-compilers` which uses `patchelf` to add the RPATHs to some binaries. You ask in https://github.com/spack/spack/pull/22322#discussion_r611606972 whether there are limitations in adding RPATHs to binaries - I think `patchelf` can do this (see also: https://stackoverflow.com/questions/13769141/can-i-change-rpath-in-an-already-compiled-binary).

Reply:
Thanks, will try to come up with a solution with `patchelf` then.
As discussed on Slack, using `LD_LIBRARY_PATH` can cause other problems, either breaking binaries from other modules or even the host binaries (since they do not use `RPATH` but rely on `/etc/ld.conf`).

Reply:
Seems to work, PR updated.

Reply:
Thanks!